### DISCUSSION TITLE
Custom App Launch Buttons

### DISCUSSION THREAD

**Discussion by PRProd** [PRProd/HA-Firemote#5] [Announcements]
- Substantial amount of custom app launch button code was added https://github.com/PRProd/HA-Firemote/discussions/4
- Added haptic feedback for button presses

<hr /><em>This discussion was created from the release <a href='https://github.com/PRProd/HA-Firemote/releases/tag/v1.5.0'>Custom App Launch Buttons</a>.</em>

**Comment by toddsay** [comment]
Thanks, this feature is really great! My votes for other app choices are: AMC+, AppleTV, Paramount+, and possibly even Kodi? (not sure about Kodi, as it's not even in the app store and has to be sideloaded... it just seems like something some would use).

Also what do you think about adding a few more buttons? I know the real remote only has 4, but it seems like there is easily space for 2 more there (and that's assuming that all the other existing buttons are actually usable, not sure if they all will be).

I successfully used Nextlix/Hulu/Disney+/HBO Max, but not Prime Video (although I can just hit the home button to get me back to the mostly-prime content). Unfortunately none of the buttons (anywhere) are working very well for me at the moment. Not sure what the issue is, as it was working at one time as a "4 Series" (it's actually a 4K Max, which is the setting I'm using now). I also noticed that my entity shows up two times in the list, but identically named (not sure if that's normal or part of the symptom). The only way I can get any buttons to work is using "Strong", and even that seems a bit inconsistent and with a few seconds delay for each press.

**Reply by PRProd** [reply]
Glad you like the changes!  I'll hopefully be able to add your app requests soon, as personal time permits.

As far as your buttons not working anymore, I think the troubleshooting solution is actually a simple one.  Under the compatibility mode option (which I'm just noticing now, I misspelled 'compatibility' in the UI.  Ha!) you may have to experiment with the various event options to find one that works best for your system.  Initially, I was informed that event1 was the winner for the 4K max, but maybe that's not the case.  Because of the info I was given 'event 1' is actually the default when you select the 4K max.  Depending on your test results, I could change the default to help others avoid this issue in the future.  I'd love to personally test them all myself, but devices aren't free and there isn't a lot of money in volunteering... so, if you could put in the leg work for me to find out which one works the best, I'd very much appreciate hearing your results.  :-)

"Strong" will always be slow, unfortunately.  That's the nature of the way those commands work, hence the "(Slower)" disclaimer I attached to that option.  The preferred way is to iron out the event number, and then any remaining broken buttons can be handled one by one as they are reported.  This will ensure a much faster response time.  As far as your entity showing up twice in the list, that's something I'm aware of and attempting to remedy.  It's minor, and it doesn't change the functionality of the card, so no need to worry there.  It's just a cosmetic thing in the setup menu.

Keep me posted, and thanks again for your help!

**Reply by toddsay** [reply]
Thanks again. So, in my case with a 4K Max, 'event 5' is the one that works. 

However, there's something else strange going on, which now explains why I was so confused: There seems to be something about my main dashboard that is causing trouble or interfering in some way. I added this card to my smaller "mobile" dashboard and it is working fine there (using event 5). But on my more complex dashboard with the same config, the button presses work only occasionally and with a delay. Most of the presses do nothing at all. I'm not sure if there is a particular card causing a conflict, or if it's just the number of cards active at one time (it's probably about 70 cards total, nested in various ways within horizontal/vertical cards).

So, yeah... weird. Not sure if you would even have thoughts on that one. I may try building alternate dashboards with subsets to try and narrow down what might be causing a conflict.

**Reply by PRProd** [reply]
I just got my hands on a 4K Max stick this afternoon, so I'm going to look into some stress testing this weekend (hopefully) to see if mine agrees with yours that "event 5" should be the default setting.

About the bigger issue, I'm curious about a few things.  Do you notice any layout shifting issues when your dashboard loads?  Things flashing, or other such nonsense?  I'd love to hear if any if your other dashboards present you with this same problem.  I've done most of my testing on dashboards with 10 or fewer cards on them, (nowhere near 70) but perhaps I should give it a try.  It wouldn't surprise me if I did something wrong (rookie mistake?)... and if so, hopefully I can correct it easily.

### Coming up in the next minor release:
 - More app launch buttons!  
   - amc+
   - AppleTV
   - Paramount+
 - A (finally) fixed stable "prime video" app launch button

### Tasks for future unscheduled releases:
 - evaluation and/or change for the default event value on Fire TV Stick 4K MAX
 - HDMI source input buttons available for physical TVs
 - Allow more than 4 custom app launch buttons on remotes that already have them
 - YAML based user button overrides for specific buttons, allowing for scripts or services to be called instead of the default action for that button (from: #6 )

**Reply by toddsay** [reply]
Thanks, that sounds great! And I did more experimentation, and finally realized my "unresponsive" problem may be more of a state the page was in. It's working well at the moment and I don't think anything should have changed (I mean, I added a new dashboard, edited, removed, etc, but the end state looks the same as the begin state). It was exhibiting the problem a bit earlier, and I learned that when it fails, it's not because it isn't transmitting, but because it isn't registering the click. That is, you can't actually see the visual pressing of the button effect. I'm not sure what causes it, but it may be something entirely outside of your control. I'll let you know if I find more, but you probably shouldn't worry about that one.

**Comment by PRProd** [comment]
Fun facts:  (well, fun for me anyway, and maybe helpful for future me or someone else in the future)

My interrogation of the Amazon Fire TV 4K MAX device shows the following inputs:
| name | path | huh? |
| ---- | ---- | ---- |
| hdmipower | /dev/input/event0 | I don't know, but I'm intrigued |
| WOBLE_INPUT_DEVICE |  /dev/input/event1 | Game Controller maybe? |
| amazon_touch | dev/input/event2 | Touchscreen controller maybe? |
| WOW_INPUT_DEVICE | /dev/input/event3 | Much wow!  Very excite!  also... I don't know, but maybe this? https://www.urcsupport.com/urc_product/wow-remote/ |
| kcmouse | /dev/input/event4 | a mouse I'd assume, or a more Spanish version of "Keyboard con Mouse" |
| Amazon Fire TV Remote | /dev/input/event5 | An expected amazon remote - The winner for my "what should the default be" question since we are trying to pretend we are a remote |

Also for future self and/or google-ers, the useful ADB command that brought me this info is
`adb shell dumpsys input`

This is also possible through the Home Assistant UI with a few steps:
 - Open Developer Tools
 - Click Services Tab
 - Select "Android TV: ADB Command" service
 - Choose your device by clicking on the entity button
 - enter `adb shell dumpsys input` into the command field
 - click the CALL SERVICE button
 - click over to the developer tools States tab
 - choose your entity (the one you sent the command to)
 - review the data to see what and how your device is listening for external input

### Conclusion
I'm going to change the default for the Amazon Fire TV 4K MAX device to event5

**Comment by PRProd** [comment]
### Release 1.5.1 available.
- Added buttons for amc+, Apple TV, and Paramount+
- Improved reliability for prime video button
- Improved reliability for 4K Stick MAX by changing default event to 5

#### Tasks for future unscheduled releases:
 - HDMI source input buttons available for physical TVs
 - Allow more than 4 custom app launch buttons on remotes that already have them
 - YAML based user button overrides for specific buttons, allowing for scripts or services to be called instead of the default action for that button (from: https://github.com/PRProd/HA-Firemote/issues/6 )

**Reply by toddsay** [reply]
Thanks, the 4k Max is working for me now with the default event setting. Looking forward to those future changes too, all great ideas! This could end up being almost a "Harmony remote" style control card, with buttons to control just about anything. After that I'm going to want even more buttons on it. 😆 

I did more experimentation on my "unresponsive" problem today. Basically there's a long delay or the button press is missed entirely. But I found out what is causing it! I've got a "mini-media-player" card on the same dashboard that is also controlling the Fire TV. Having both seems to cause a conflict (removing that card restores the expected behavior). Not sure if that is fixable, or just a reasonable limitation.

**Comment by PRProd** [comment]
My interrogation of the **Amazon Fire TV Stick Lite** device shows the following inputs:
| name | path | huh? |
| ---- | ---- | ---- |
| hdmipower | /dev/input/event0 | I don't know |
| WOBLE_INPUT_DEVICE |  /dev/input/event1 | Game Controller maybe? |
| amazon_touch | dev/input/event2 | Touchscreen controller maybe? |
| kcmouse | /dev/input/event3 | a mouse I'd assume |
| Amazon Fire TV Remote | /dev/input/event4 | A physical remote control that appears only when a remote is currently attached/associated with this device |

Because of this, the default option when a Fire TV Stick Lite device is chosen is **event4**

<br><br>
My interrogation of the **Amazon Fire TV Stick 4K** device shows the following inputs:
| name | path | huh? |
| ---- | ---- | ---- |
| hdmipower | /dev/input/event0 | I don't know |
| WOBLE_INPUT_DEVICE |  /dev/input/event1 | Game Controller maybe? |
| amazon_touch | dev/input/event2 | Touchscreen controller maybe? |
| kcmouse | /dev/input/event3 | a mouse I'd assume |
| Amazon Fire TV Remote | /dev/input/event4 | A physical remote control that appears only when a remote is currently attached/associated with this device |

Because of this, the default option when a Fire TV Stick 4K device is chosen is **event4**
<br><br>

My interrogation of the **Amazon Fire TV (4 Series)** device shows the following inputs:
| name | path | huh? |
| ---- | ---- | ---- |
| MStar Smart TV IR Receiver | /dev/input/event0 | Physical IR remote receiver |
| MTK TV KEYPAD |  /dev/input/event1 | I don't know |
| amazon_touch | dev/input/event2 | Touchscreen controller maybe? |
| kcmouse | /dev/input/event3 | a mouse I'd assume |
| amznkeyboard | /dev/input/event4 | an amazon keyboard? |
| Amazon Fire TV Remote | /dev/input/event5 | A physical remote control that appears only when a remote is currently attached/associated with this device |

Because of this, the default option when a Fire TV (4 Series) device is chosen is **event0**

**Comment by PRProd** [comment]
@toddsay - Version 1.6.0 has been released!  Feel free to [join the discussion here](https://github.com/PRProd/HA-Firemote/discussions/8).  
This release allows for 2 additional app launch buttons on your remote, plus other stuff.  Enjoy!

**Comment by toddsay** [comment]
Thanks! I noticed, and am already enjoying the changes. Nice work!

**Comment by Scoff123** [comment]
Thank you for this card and all of your hard work!! 
Is it at all possible for the BBC iPlayer app to be a custom button launch for the 4k Max?

**Reply by PRProd** [reply]
Maybe it's because I'm in the US, but when I try installing the "BBC iPlayer" app on my device, the download page says "Unavailable".  Without being able to download it for myself, it might take a while to get this added.  Would you be willing to help me with some testing?

If we are lucky, I'll only need you to do this:
1. Launch the app on your device
2. In Home Assistant, click on Developer Tools -> States
3. Find your 4K Max entity by scrolling, or clicking the "set state" bar, and searching for it
4. Reply to this post with the contents found in the Attributes column or the "State attributes (YAML, optional)" box

**Reply by Scoff123** [reply]
Thanks for the quick response! Happy to help with this 😊 
We may be in luck, I've done as you suggested and here are the Attributes:

adb_response: >-
  Starting: Intent { act=android.intent.action.MAIN
  cat=[android.intent.category.LAUNCHER]
  cmp=com.amazon.firebat/.deeplink.DeepLinkRoutingActivity }
hdmi_input: null
device_class: tv
friendly_name: Fire TV 192.168.1.135
supported_features: 22961
source_list:
  - uk.co.bbc.iplayer
app_id: uk.co.bbc.iplayer
source: uk.co.bbc.iplayer
entity_picture: >-
  /api/media_player_proxy/media_player.fire_tv_192_168_1_135?token=c0a7d136150a117438fa4de0f3d181beddbccf040489f6758a6c36966d56610d&cache=1668160447.442522

**Reply by PRProd** [reply]
@Scoff123 - Release [v1.8.0](https://github.com/PRProd/HA-Firemote/releases/tag/v1.8.0) is ready, which includes an untested BBC iPlayer App shortcut.  Could I ask you to update to the newest Firemote version, and let me know if the app shortcut button works or not?  That would be very helpful.  Thank you.

**Reply by Scoff123** [reply]
@PRProd - This is great, I've updated to 1.8.0, added BBC iPlayer as one of my 4 buttons and it works perfectly - thank you!! 😊
